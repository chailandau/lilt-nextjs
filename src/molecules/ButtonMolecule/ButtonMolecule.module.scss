.button {
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: rem(20);
    padding: rem(12) rem(20);
    @include font(14, 16);
    @include font-rubik-bold;
    color: color(neutral, black);
    text-transform: uppercase;
    flex-grow: 1;
    transition: flex-grow 0.3s ease-in-out;
    @media (prefers-reduced-motion) {
        transition: none;
    }
    @include hover {
        color: color(neutral, black);
    }
    @include tablet {
        flex-grow: 0;
    }
    &.blue {
        @include bgSlideHover(
            color(blueTheme, neonBlue),
            color(neutral, white)
        );
    }
    &.green {
        @include bgSlideHover(
            color(blueTheme, neonGreen),
            color(neutral, white)
        );
    }
    &.white {
        @include bgSlideHover(
            color(neutral, white),
            color(blueTheme, neonBlue)
        );
    }
    &.link {
        @include font-rubik-regular;
        @include font(16, 20);
        color: color(neutral, white);
        text-transform: uppercase;
        padding: rem(14);
        margin: rem(-14);
        @include underlineSlideHover;
    }
    &.green-theme {
        &.blue {
            @include bgSlideHover(
                color(greenTheme, neonBlue),
                color(neutral, white)
            );
        }
        &.green {
            @include bgSlideHover(
                color(greenTheme, neonGreen),
                color(neutral, white)
            );
        }
        &.white {
            @include bgSlideHover(
                color(neutral, white),
                color(greenTheme, neonBlue)
            );
        }
    }
}
