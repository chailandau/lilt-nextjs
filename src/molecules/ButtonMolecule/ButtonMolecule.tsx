'use client';

import classNames from 'classnames';
import { FC, KeyboardEvent, ReactNode } from 'react';

import styles from './ButtonMolecule.module.scss';

import Link from '@/atoms/Link';

export const buttonMoleculeVariants = [
    'blue',
    'green',
    'white',
    'link'
] as const;

export type ButtonMoleculeVariant = (typeof buttonMoleculeVariants)[number];

export interface ButtonMoleculeProps {
    /** Button variant */
    variant?: ButtonMoleculeVariant;
    /** Button content */
    children: ReactNode;
    /** `Button` renders as an `a` tag if supplied */
    link?: string;
    /** `onClick` event */
    onClick?: () => void;
    /** `onKeyDown` event */
    onKeyDown?: (e: KeyboardEvent<HTMLButtonElement>) => void;
    /** CSS `className` */
    className?: string;
}

const ButtonMolecule: FC<ButtonMoleculeProps> = ({
    variant = 'blue',
    link,
    onClick,
    onKeyDown,
    children,
    className
}) => {
    const classList = classNames(
        styles['button'],
        styles[variant],
        className && className,
        styles[process.env.NEXT_PUBLIC_THEME as string]
    );

    const isLink = link && link.trim() !== '';

    return isLink ? (
        <Link
            href={link}
            className={classList}
            underlineHover={false}
            onClick={onClick}
        >
            {children}
        </Link>
    ) : (
        <button
            className={classList}
            onClick={onClick}
            onKeyDown={(e) => onKeyDown && onKeyDown(e)}
        >
            {children}
        </button>
    );
};

export default ButtonMolecule;
