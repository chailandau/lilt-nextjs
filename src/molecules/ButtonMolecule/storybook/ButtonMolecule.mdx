import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as ButtonMoleculeStories from './ButtonMolecule.stories';

<Meta of={ButtonMoleculeStories} />

# Button Molecule

<Canvas of={ButtonMoleculeStories.Default} />

<br />

## children

Text is supplied as a child to the `Button` molecule.

## color

Controls the color of the `Button` molecule.

## link

If the link prop is not empty, the `Button` will render as an `a` tag. Otherwise, it will default to a `button` tag.

## onClick

If there is no link and the `onClick` prop is supplied, it will be called when the `Button` is clicked.

<br />
<br />
<br />

## Props Table

<ArgTypes />
