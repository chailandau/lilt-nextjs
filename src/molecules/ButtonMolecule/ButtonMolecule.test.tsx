import { composeStory } from '@storybook/react';
import { render, screen } from '@testing-library/react';

import { buttonMoleculeVariants } from './ButtonMolecule';
import Meta, { Default } from './storybook/ButtonMolecule.stories';

import {
    testAppliesClass,
    testAxeAndSnapshot,
    testKeyDown,
    testPropOptions,
    testRenderText
} from '@/utils/testHelpers';

const ButtonMolecule = composeStory(Default, Meta);

describe('Button Molecule', () => {
    const buttonMolecule = {
        component: <ButtonMolecule />
    };
    testRenderText({
        ...buttonMolecule,
        role: 'button',
        text: ButtonMolecule.args.children as string
    });
    it('has default color', () => {
        render(<ButtonMolecule variant={undefined} />);
        expect(screen.getByRole('button')).toHaveClass('blue');
    });
    testPropOptions({
        component: <ButtonMolecule />,
        propName: 'variant',
        propOptions: buttonMoleculeVariants,
        htmlTag: 'button'
    });
    it("renders 'link' correctly", () => {
        render(<ButtonMolecule link='https://google.com' />);
        expect(screen.getByRole('link').getAttribute('href')).toBe(
            'https://google.com'
        );
    });
    it("renders 'button' correctly", () => {
        render(<ButtonMolecule />);
        expect(screen.getByRole('button')).toBeDefined();
    });
    testKeyDown({ ...buttonMolecule, role: 'button' });
    testAppliesClass({ ...buttonMolecule, htmlTag: 'button' });
    testAxeAndSnapshot(buttonMolecule);
});
