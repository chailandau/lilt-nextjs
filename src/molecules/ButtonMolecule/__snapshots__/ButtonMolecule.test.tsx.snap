// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Molecule matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex"
  >
    <button
      class="button blue"
    >
      Default button
    </button>
  </div>
</DocumentFragment>
`;
